### ISSUE TITLE
associated header sometimes not recognized

### ISSUE THREAD

**Issue by moschny** [include-what-you-use/include-what-you-use#279]
iwyu has the notion of an "associated" header file. So when there are `foo.c` and `foo.h` in the same directory, iwyu detects that they belong together, and puts an `#include "foo.h"` at the beginning of `foo.c`. Also, it treats all `#include` lines in `foo.h` as if they were also present in `foo.c`, which is nice.

Now, in our project, we have `foo.h` not in the same directory as foo.c but in a separate `include/` directory. iwyu treats `foo.h` as a "normal" header file in this case.

Is it possible to instruct iwyu to treat `foo.h` as associated with `foo.c` even when they are not in the same directory?

If not it would be nice to add such a feature.

**Comment by kimgr** [comment]
We have the same setup, and even include the associated header with a prefix, so e.g.

    include/
        xyz/
            foo.h
    source/
        foo.cpp

    // foo.cpp
    #include "xyz/foo.h"

The path to `include` is in our `-I` when analyzing `foo.cpp`. This works for us. How is your setup different?

**Comment by moschny** [comment]
Did not yet find a time to create a small example to actually prove it, but this is my current theory: iwyu 
recognizes `foo.h` as associated for `foo.c` iff one (or both) of these conditions apply:
* `#include "foo.h"` is the first include directive in `foo.c`
* `foo.h` and `foo.c` live in the same directory

If neither is true, the association doesn't seem to be recognized.

**Comment by kimgr** [comment]
Yes, I can confirm that I've seen IWYU fail to recognize the associated header if it's not the first `#include` in the source file. I've just manually moved it up in those cases, and run the IWYU analysis again.

**Comment by moschny** [comment]
Ok, thanks for the confirmation.

So, I'd consider this a bug.

**Comment by kimgr** [comment]
I'm worried that that would weaken the heuristic too much, see the rule implemented here:
https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu_preprocessor.cc#L964

Then again, I'm not sure if ignoring the file's `#include` index would cause any trouble in practice... Thoughts?

**Comment by moschny** [comment]
Maybe one could add a switch to opt-out from the purely name-based heuristic?

As an additional data point, `fix_includes -s` seems to always move `#include foo.h` to the top, without knowing the path to `foo.h`.

**Comment by vsapsai** [comment]
To change the heuristic we need some data. I was thinking about checking certain number of C++ projects on GitHub and seeing how many non-unique file names are out there. For example, if every project has `foo.h` then it is not safe to assume in `foo.cc` that any encountered `foo.h` is associated header.

Another thing to consider is what are consequences of false negatives and false positives. If incorrectly treating a file as associated header causes more troubles than missing associated header detection, we'd better keep heuristic strict.

**Comment by Flamefire** [comment]
Current heuristics: If path and name (w/o extension) match, it is associated. If not it checks if the first (or in case of a pch he 2nd) file has the same basename and marks this as associated. This will make `include/bar/foo.h` associated to ` foo.cc` if the former is the first include line in the latter.

Note that the heuristic fails currently, when the cc-file is given by its absolute path (part of fix in #276)

I would make a slight change to the 2nd part of the heuristic: Check the basename of the first include always (no need for the pch-in-code option). If the first is not the associated include, do the same to the 2nd. This allows usage of pchs in some part of the code (e.g. multi-library CMake project where one library has a pch and the other has not)

**Comment by kimgr** [comment]
There's now an `// IWYU pragma: associated`, introduced in IWYU 0.11, that can be put next to any `#include` statement.

Closing as fixed.
